### PR TITLE
#10322: Fix scale selector to make it compatible with different projections

### PR DESCRIPTION
**Pull request by mahmoudadel54**
## Description
In this PR, the issue of scale selector which was not compatible with different projections is resolved.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [x] Bugfix
 - [ ] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

## Issue
#10322

**What is the current behavior?**
#10322 

**What is the new behavior?**
Now if user opens the print modal, switches the projection to 4326 for example and select a scale from the scale selector list, it will selected successfully without choosing wrong one.

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

## Other useful information
I have noticed some points within this issue: 

1- wrong calculate scale zoom level in case user changes projection in the print modal, this causes the noticed wrong calculated scale in DD scales. 
If you see this line which calculate the scaleZoom if useFixedScales = true 
https://github.com/geosolutions-it/MapStore2/blob/7443a4839e8fed1f1cd981a578b404a7e2ccde55/web/client/plugins/Print.jsx#L593 
it uses getNearestZoom method: https://github.com/geosolutions-it/MapStore2/blob/7443a4839e8fed1f1cd981a578b404a7e2ccde55/web/client/utils/PrintUtils.js#L187-L192 and into this method it uses defaultScales is applied at each projection and this defaultScales is the google mercator scales : https://github.com/geosolutions-it/MapStore2/blob/7443a4839e8fed1f1cd981a578b404a7e2ccde55/web/client/utils/PrintUtils.js#L42

So what I did is passing the scales of the actual correct projection to this method to get the correct calculated scaleZoom to display the correct expected scale in the scales drop down list.

2- Based on the 1st point above, the scale value of the configure print map was calculated wrongly: https://github.com/geosolutions-it/MapStore2/blob/7443a4839e8fed1f1cd981a578b404a7e2ccde55/web/client/plugins/Print.jsx#L594. Another point I need to mention is that mapZoom of the map viewer is passed to **configurePrintMapProp** instead of the calculated zoom value of the print map based on the selected projection here: 
https://github.com/geosolutions-it/MapStore2/blob/7443a4839e8fed1f1cd981a578b404a7e2ccde55/web/client/plugins/Print.jsx#L596 and here as well 
https://github.com/geosolutions-it/MapStore2/blob/7443a4839e8fed1f1cd981a578b404a7e2ccde55/web/client/plugins/Print.jsx#L599

so here I refactored the method to calculate needed values based on the projection value

3- using '**getZoomForExtent**' in '**configurePrintMap**' method is not the proper way to get the mapZoom of the print map --> in some cases the change bet. main map viewer and the print map viewer >= 2 and this makes change in the print result: https://github.com/geosolutions-it/MapStore2/blob/7443a4839e8fed1f1cd981a578b404a7e2ccde55/web/client/plugins/Print.jsx#L583-L592

so I used reprojectZoom method to calculate print map zoom instead using bbox and mapSize for that: https://github.com/geosolutions-it/MapStore2/blob/138ae35d89e9cf124314e41602cb3cf39a731a8f/web/client/plugins/Print.jsx#L586.

4- when user opens the print panel, the sent projection 'prop' should be set initially to be consistent with the UI [selected projection in DD] and to be not misleading if the shown projection into UI not the initial one.

So I used initiated the projection into useEffect in Projection component at initiating the component: 
https://github.com/geosolutions-it/MapStore2/blob/138ae35d89e9cf124314e41602cb3cf39a731a8f/web/client/plugins/print/Projection.jsx#L68-L75

### CODE REVIEW

**Comment by tdipisa** [comment]
@offtherailz I would like if you can schedule a call with @mahmoudadel54 to review issues requirements before moving on with this review.

**Review by offtherailz — CHANGES_REQUESTED** [review]
- [ ] I  noticed that the printing result is still wrong when switching coordinates. It have to be consistent with the map. 

[screencast-localhost_8081-2024.06.18-18_21_32.webm](https://github.com/geosolutions-it/MapStore2/assets/1279510/32493050-c870-4455-85d1-b55db021b146)

Everything have to be consistent. I'll review the code where it is fixed. 

- [ ] Other useful information are not clear, please explain them better.

File: web/client/plugins/__tests__/Print-test.jsx [hunk 1]
```diff
@@ -202,6 +202,88 @@ describe('Print Plugin', () => {
...
+        };
+        getPrintPlugin({
+            state: {...initialState,
+                print: {...initialState.print,
+                    capabilities: {...initialState.print.capabilities,
+                        scales: [1000000, 500000, 100000].map(value => ({name: value, value}))}
+                }}
+        }).then(async({ Plugin }) => {
+            try {
+                await ReactDOM.render(<Plugin
```
Comment (offtherailz) [inline web/client/plugins/__tests__/Print-test.jsx:227]: I don't think ReactDOM.render returns a promise

File: web/client/plugins/__tests__/Print-test.jsx [hunk 2]
```diff
@@ -202,6 +202,88 @@ describe('Print Plugin', () => {
...
```
Comment (offtherailz) [inline web/client/plugins/__tests__/Print-test.jsx:268]: same here. Await makes sense only on promises.

File: web/client/plugins/__tests__/print/Projection-test.jsx [hunk 3]
```diff
@@ -119,11 +120,12 @@ describe('PrintProjection Plugin', () => {
...
     it('map transformer with user chosen crs', (done) => {
-        getPrintProjectionPlugin().then(({ Plugin, store }) => {
+        getPrintProjectionPlugin().then(async({ Plugin, store }) => {
             try {
-                ReactDOM.render(<Plugin projections={[{"name": "Mercator", "value": "EPSG:3857"}, {"name": "WGS84", "value": "EPSG:4326"}]}/>, document.getElementById("container"));
+                const comp = ReactDOM.render(<Plugin projections={[{"name": "Mercator", "value": "EPSG:3857"}, {"name": "WGS84", "value": "EPSG:4326"}]}/>, document.getElementById("container"));
+                await ReactTestUtils.act(async() => comp);
                 const combo = getByXPath("//select");
-                ReactTestUtils.Simulate.change(combo, {
+                await ReactTestUtils.Simulate.change(combo, {
```
Comment (offtherailz) [inline web/client/plugins/__tests__/print/Projection-test.jsx:128]: `.Simulate.change` does not return a promise, so using `await` is wrong.

File: web/client/plugins/__tests__/print/Projection-test.jsx [hunk 4]
```diff
@@ -139,11 +141,12 @@ describe('PrintProjection Plugin', () => {
...
     it('validator without allowPreview', (done) => {
-        getPrintProjectionPlugin().then(({ Plugin, store }) => {
+        getPrintProjectionPlugin().then(async({ Plugin, store }) => {
             try {
-                ReactDOM.render(<Plugin projections={[{"name": "Mercator", "value": "EPSG:3857"}, {"name": "WGS84", "value": "EPSG:4326"}]}/>, document.getElementById("container"));
+                const comp = ReactDOM.render(<Plugin projections={[{"name": "Mercator", "value": "EPSG:3857"}, {"name": "WGS84", "value": "EPSG:4326"}]}/>, document.getElementById("container"));
+                await ReactTestUtils.act(async() => comp);
                 const combo = getByXPath("//select");
-                ReactTestUtils.Simulate.change(combo, {
+                await ReactTestUtils.Simulate.change(combo, {
```
Comment (offtherailz) [inline web/client/plugins/__tests__/print/Projection-test.jsx:149]: .Simulate.change does not return a promise, so using await is wrong.

**Comment by mahmoudadel54** [comment]
> I noticed that the printing result is still wrong when switching coordinates. It have to be consistent with the map. [screencast-localhost_8081-2024.06.18-18_21_32.webm](https://github.com/geosolutions-it/MapStore2/assets/1279510/32493050-c870-4455-85d1-b55db021b146)
> 
> Everything have to be consistent. I'll review the code where it is fixed.

Hi @offtherailz 

After checking this issue, I found out, this is the same currently behavior in print [in DEV or QA] ---> if the print projection is one of the geographic coordinate systems [like 4326] and the **'geodetic'** = true  into the "overrideOptions" in print cfg, and when I edited **'geodetic'** to be false, printing map with geographic projection e.g: 4326 printing works for me.
Honestly I don't know the reason.
This is a demo of my debugging: 

https://github.com/geosolutions-it/MapStore2/assets/58145645/b0084bac-1579-4742-b888-db22244b1863

**Comment by mahmoudadel54** [comment]
> * [ ]  Other useful information are not clear, please explain them better.

@offtherailz  I have updated the useful information section. I hope it is more clear now

**Review by offtherailz — CHANGES_REQUESTED** [review]
AS we discussed in a call @mahmoudadel54 the PR has still the inonsistencies I notified in the Preview.

As we checked, the preview looks consistent when `geodetic` is set to `false`, so our development will do it in that direction, assuming that `geodetic` is set to `false`.
Now print plugin allows to send the geodetic parameter via configuration, but it doesn't handle to change the preview to fit the final result.

**Comment by offtherailz** [comment]
Now it looks a lot better.
I'm reporting here in a comment my findings just to keep track of them.
I'll check if these issues of the preview already existed and let you know.

Discrepancy between the preview and the scale sent (as we have seen in our meeting).

image

**Review by offtherailz — CHANGES_REQUESTED** [review]
Please fix the tests and do a sert of manual tests more.

**Review by offtherailz — APPROVED** [review]

**Comment by tdipisa** [comment]
@ElenaGallo please test this in DEV once deployed and let us know for the backport.

**Comment by ElenaGallo** [comment]
Test passed, @mahmoudadel54 please backport to 2024.01.xx Thanks

**Comment by mahmoudadel54** [comment]
> Test passed, @mahmoudadel54 please backport to 2024.01.xx Thanks

@ElenaGallo Backport is done ---> https://github.com/geosolutions-it/MapStore2/pull/10452

**Comment by tdipisa** [comment]
@ElenaGallo I've also tested using test map [here](https://dev-mapstore.geosolutionsgroup.com/mapstore/#/context/demo_context/49601) and all seems to work as expected with this fix.

**Pull request merged**